flow/damage_context.h

```diff
@@ -0,0 +1,106 @@
+#ifndef FLUTTER_FLOW_CONTEXT_H_
+#define FLUTTER_FLOW_CONTEXT_H_
+
+#include <unordered_set>
+#include "flutter/flow/embedded_views.h"
+#include "third_party/skia/include/core/SkRect.h"
+
+class SkImageFilter;
+
+namespace flutter {
+
+struct PrerollContext;
+class Layer;
+
+class DamageArea {
+ public:
+  FML_DISALLOW_COPY_AND_ASSIGN(DamageArea);
+
+  DamageArea() = default;
+  DamageArea(DamageArea&&) = default;
+  DamageArea& operator=(DamageArea&&) = default;
+
+  const SkIRect& bounds() const { return bounds_; }
+
+  std::vector<SkIRect> GetRects() const;
+
+  void AddRect(const SkRect& rect);
+  void AddRect(const SkIRect& rect);
+
+ private:
+  SkIRect bounds_ = SkIRect::MakeEmpty();
+};
+
+class DamageContext {
+ public:
+  typedef bool (*LayerComparator)(const Layer* l1, const Layer* l2);
+
+  // Opaque representation of a frame contents
+  class FrameDescription;
+
+  void InitFrame(const SkISize& frame_size,
+                 const FrameDescription* previous_frame_description);
+
+  void PushLayerContribution(const Layer* layer,
+                             LayerComparator comparator,
+                             const SkMatrix& matrix,
+                             const PrerollContext& preroll_context,
+                             size_t index = size_t(-1));
+
+  size_t layer_entries_count() const { return layer_entries_.size(); }
+
+  // Adjusts the paint bounds of layer entries within given range according to
+  // filter bounds
+  bool ApplyImageFilter(size_t from,
+                        size_t count,
+                        const SkImageFilter* filter,
+                        const SkMatrix& matrix,
+                        const SkRect& bounds);
+
+  struct DamageResult {
+    DamageArea area;
+    std::unique_ptr<FrameDescription> frame_description;
+  };
+
+  DamageResult FinishFrame();
+
+ private:
+  // Represents a layer contribution to screen contents
+  // LayerContribution can compare itself with a LayerContribution from past
+  // frame to determine if the content they'd produce is identical
+  // Diffing set of LayerContributions will give us damage area
+  struct LayerContribution {
+    SkRect paint_bounds;  // in screen coordinates
+    std::shared_ptr<const Layer> layer;
+    LayerComparator comparator;
+
+    std::vector<std::shared_ptr<Mutator>> mutators;
```

Comment:
If the DamageContext (and/or the PrerollContext) kept a linked list of Mutator objects then that would save a lot of vector creation of vectors that mostly contain the same objects.

A LayerContribution would simply copy over the pointer from the Context at the time of its creation.

Adding a mutator would simply be "mutator_links = MutatorLink(Mutator, mutator_links)"
Removing a mutator would simply be "mutator_links = mutator_links.next"

Comparisons could even be short circuited. Once we determine that linkX from the previous frame matches linkY from the new frame, then we can tag them as being equal (which implies that their "next" chains are also equal) then we can tag them as equal and avoid recursive comparisons when they appear opposite each other on future chain tests.

Reply:
I remember this from first suggestion. I like the idea, but I think the list would need to be double linked (mutator stack exposes both iterator and reverse_iterator). And the iterators would also need to be implemented.

Reply:
Double linking the list would remove the ability to snapshot it with a pointer copy. The idea was that the earlier entries on the chain, coming from the top-level layers, would stay there and occasionally have a new mutator point to them. But if they also need a reverse pointer then we run into a problem.

The only place I saw iteration was in the comparison of the list. Is that any simpler than double-side-by-side iterations of the two linked lists instead?

Reply:
The stack seems to be iterated from bottom to top in `EmbedderLayers::PushPlatformViewLayer` and in `FuchsiaExternalViewEmbedder::BeginFrame`. I mean technically even with double linked the 'prev' pointer (top to bottom) never changes, so we can "snapshot" the top mutator, as long as we ignore the other pointer and only iterate to the bottom, which is enough for damage_context.

Reply:
Ah, yes, if we replace the stack in the PrerollContext, then that would be an issue. If we shadow the stack in the DamageContext then we can get away with it. Since we are currently shadowing the stack in every Contribution object then shadowing the stack once in DamageContext would still reduce storage requirements by a large margin.

I think you are correct that snapshotting would still work if the snapshot list is only iterated forward, but I need to mull it a little more.

Also, only doing the linked list in a shadow copy means we don't need to modify the entire engine. Perhaps we could do that for starters and if this gets in there and gets vetted then we can have a follow-on task to promote that to the PrerollContext.

Reply:
Possible solution: https://github.com/knopp/engine/commit/3179ade95c59ad72b00b5611b89adedde847c210#diff-7b84dd247cd3f78e219d26495e874ef4

It doesn't disrupt existing code in any way and reference to `MutatorNode` is essentially snapshot of `MutatorStack`.